parser/parser_reader.js

```diff
@@ -0,0 +1,73 @@
+"use strict";
+
+const { EOF } = require("../constants/globals.js");
+
+/* Tests if a character is a whitespace character */
+const isSpace = c => /\s/.test(c);
+
+/* Tests if a character is a tag delimeter */
+const isTagDelim = c => (c === '/' || c === '>');
+
+/* Tests is a character is EOF */
+const isEOF = c => (c === EOF);
+
+/* Currently this allows html tags to contain any characters that aren't spaces */
+const isValidTagCharacter = c => !(isEOF(c) || isTagDelim(c) || isSpace(c));
+
+/* Tests if a character is an english alphabetical character */
+const isAlpha = c => /[a-zA-Z]/.test(c)
+
+function Reader(file) {
+  this.file = file;
+  this.index = 0;
+}
+
+/* Increments reader's index to next non-space character */
+Reader.prototype.trimWhiteSpace = function() {
+  while((this.index < this.file.length) && isSpace(this.file[this.index])) { this.index++; }
+}
+/* Return next character in file stream and increments the Reader's index */
+Reader.prototype.nextChar = function() {
+  if(this.index >= this.file.length) {
+    return EOF;
+  }
+  return this.file[this.index++];
+};
+
+/* Returns next character in file without advancing Reader's index */
+Reader.prototype.peekChar = function() {
+  if(this.index >= this.file.length) {
+    return EOF;
+  }
+  return this.file[this.index];
+};
+
+/* Returns next non-space character */
+Reader.prototype.next = function() {
+  this.trimWhiteSpace();
+  return this.nextChar();
+};
+
+/* Returns next non-space character in file without advancing Reader's index */
+Reader.prototype.peek = function() {
+  let peekIndex = this.index;
+  while((peekIndex < this.file.length) && isSpace(this.file[peekIndex])) { peekIndex++; }
+  if(peekIndex >= this.file.length) {
+    return EOF;
+  }
+  return this.file[peekIndex];
+};
+
+/* Returns the next HTML tag as a string. (Reads next word until space character or closing tag charcters found) */
+Reader.prototype.getTag = function() {
+  this.trimWhiteSpace();
+  let tag = "";
+  while(isValidTagCharacter(this.peekChar())) {
+    tag += this.nextChar();
+  }
+  return tag;
+};
+
+
+exports.Reader = Reader;
+exports.createParserReaders = files => files.map(file => new Reader(file));
```

Comment:
is this only to be used by the parser? Do we pass a Reader to the Parser or does the Parser call the Reader internally?

I would argue the latter, and make it independent.

At the very least, you never call it a `ParserReader` anywhere else.

Reply:
Yeah I'm working on the first part of the parser in a separate branch that I was explaining to VB above. The Reader will be in its own file and the createParserReaders can be within the parser.js file.

For every file we create a new Reader and then will pass each of those readers one by one into the parser.